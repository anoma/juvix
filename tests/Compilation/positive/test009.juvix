-- tail recursion
module test009;

open import Stdlib.Prelude;

sum' : Nat → Nat → Nat;
sum' acc zero := acc;
sum' acc (suc x) := sum' (suc x + acc) x;

sum : Nat → Nat;
sum := sum' 0;

fact' : Nat → Nat → Nat;
fact' acc zero := acc;
fact' acc (suc x) := fact' (acc * suc x) x;

fact : Nat → Nat;
fact := fact' 1;

main : IO;
main :=
    printNatLn (sum 10000) >>
    printNatLn (fact 5) >>
    printNatLn (fact 10) >>
    printNatLn (fact 12);

end;
