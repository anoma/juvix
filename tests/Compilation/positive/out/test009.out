50005000
120
3628800
479001600
